### PR TITLE
Documentation: Add a document about the purposes of DevTmpFS

### PR DESCRIPTION
**Pull request by supercomputer7**
Another document, to explain why we have a DevTmpFS, what is expected to be implemented and what are the implementation rules we want in it.

### CODE REVIEW

File: Documentation/Kernel/DevTmpFS.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# DevTmpFS and its purposes
+
+`DevTmpFS` is the current filesystem to hold device nodes in /dev directory.
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:3]:
```suggestion
`DevTmpFS` is the current filesystem to hold device nodes in the /dev directory.
```

File: Documentation/Kernel/DevTmpFS.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+so, he had to ensure the build script includes `/dev/hdb` device node, and if
+that harddrive had a partition on it, to also include `/dev/hdb1`.
+
+This solution was good for its time, serving us for a long time, until it was
+figured that this is not feasible nor the right way to expand in the future.
+
+### DevFS - the first solution for automatic device nodes
+
+`DevFS` was the first step into making the process of creating device nodes 
+automatically. The basic idea is to mount a filesystem in /dev directory, and
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:27]:
```suggestion
`DevFS` was the first step into making the process of creating device nodes 
automated. The basic idea is to mount a filesystem in /dev directory, and
```

File: Documentation/Kernel/DevTmpFS.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+to let it to handle device nodes for you. Also, you store all device nodes in
+RAM, and the build image script doesn't need to create device nodes for you
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:29]:
```suggestion
to let it to handle device nodes for you. As a result you store all device nodes in
RAM, meaning the build image script doesn't need to create device nodes for you 
```

File: Documentation/Kernel/DevTmpFS.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+automatically. The basic idea is to mount a filesystem in /dev directory, and
+to let it to handle device nodes for you. Also, you store all device nodes in
+RAM, and the build image script doesn't need to create device nodes for you 
+anymore, which was a great thing at the moment. However, due to limitations in
+the `DevFS` architecture - hardcoding names, hardcoding permissions for device 
+nodes and the lack of extensibility for future improvements in hotplug 
+capabilities in the Kernel, it was a short-lived solution and was deprecated for
+the next solution - `DevTmpFS`.
+
+As a side note, Linux used `DevFS` also in the past, but deprecated it for the
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:36]:
```suggestion
As a side note, Linux also used `DevFS` in the past, but deprecated it for the 
```

File: Documentation/Kernel/DevTmpFS.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+to let it to handle device nodes for you. Also, you store all device nodes in
+RAM, and the build image script doesn't need to create device nodes for you 
+anymore, which was a great thing at the moment. However, due to limitations in
+the `DevFS` architecture - hardcoding names, hardcoding permissions for device 
+nodes and the lack of extensibility for future improvements in hotplug 
+capabilities in the Kernel, it was a short-lived solution and was deprecated for
+the next solution - `DevTmpFS`.
+
+As a side note, Linux used `DevFS` also in the past, but deprecated it for the 
+same reason we do it in Serenity.
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:37]:
```suggestion
same reason we did it in Serenity.
```

File: Documentation/Kernel/DevTmpFS.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+There are 2 rules `DevTmpFS` has to follow:
+
+1. `DevTmpFS` doesn't assume structure of filesystem, names, or paths at all. 
+This means we don't hardcode anything in the implementation.
+2. Only device nodes, symbolic links and directories are allowed to be created.
+In contrast to `TmpFS`, we don't want unnecessary mess in this filesystem, and
+trying to create anything from userspace that doesn't match this rule will fail.
+
+Because nothing is created during the boot process (`DevFS` did that, `DevTmpFS`
+doesn't), it is the userland responsibility to generate the required device
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:64]:
```suggestion
doesn't), it is the userland's responsibility to generate the required device
```

File: Documentation/Kernel/DevTmpFS.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+nodes. How to create these device nodes is up to a userland program to decide,
+the conventional method is to look into `/sys/dev/block` and `/sys/dev/char`
+directories (which are created in the `SysFS` instance), and to look into major
+and minor numbers (which are in the layout of the filename - "major:minor"),
+then userspace can figure out based on this information which device nodes can 
+be created in the /dev directory.
+However, userspace can do that otherwise by using a statically-defined generate
+script or a list that was prepared beforehand.
```
Comment (IdanHo) [inline Documentation/Kernel/DevTmpFS.md:72]:
```suggestion
Note, however, that userspace can do so by using a statically-defined generate
script or a list that was prepared beforehand.
```

**Review by IdanHo — COMMENTED** [review]
Just a couple of grammar fixes, I'm not sure if introducing documentation about how we did things in the past is useful, but :shrug:

**Comment by supercomputer7** [comment]
> Just a couple of grammar fixes, I'm not sure if introducing documentation about how we did things in the past is useful, but shrug

I fixed the document according to the suggestions you wrote. Personally, I think that writing documentation to explain the process and what we did is a great way to explain to other Kernel developers why we came to the conclusion we need the `DevTmpFS`. Writing a document strictly about the implementation rules, and adding a bit more technical info on it, without the background, feels to me like not the right thing, because the inexperienced developer might not understand the reasons to come up with `DevTmpFS`. Also, this document is probably not a final draft, and I might add more technical info on it, just to clarify more bolts and bits of the design choices that were made when I wrote it. Of course, more changes to come are in future PRs, as always :)
Still, I've got lots of other worries in regard to `DevTmpFS`, hotplug capabilities in the Kernel and related things to it, so it will probably wait a bit when I have more time to dedicate for writing more into this document (or when I just have free time and want a context switch from all other work I do).

**Comment by awesomekling** [comment]
> I'm not sure if introducing documentation about how we did things in the past is useful

I agree with Idan. We should document how the system works today. git history keeps track of how it worked in the past :)

**Pull request closed without merging**